### ISSUE TITLE
Rule: Virtual Types should have namespaced names

### ISSUE THREAD

**Issue by jissereitsma** [extdn/extdn-phpcs#9]
Virtual Types declared in `di.xml` files should have a PSR-compliant namespaced name.

**Comment by avstudnitz** [comment]
Debatable. It leeds to thinking that they are classes which is misleading.

**Comment by jissereitsma** [comment]
You mean, just like factories and proxies. But indeed, debatable.

**Comment by renttek** [comment]
What about using an „Virtual“ prefix/suffix?
Or something like \Vendor\Module\Virtual\... ?

**Comment by jissereitsma** [comment]
I would second this entirely. I think a classname ending with `\Virtual` (similar to `\Proxy`) would be most favorable.

**Comment by schmengler** [comment]
> Debatable. It leeds to thinking that they are classes which is misleading.

there was a good discussion on the Magento Devblog about this a few months ago. IIRC a solution that found most agreement was to use PSR style class names, but with clear indication that it's virtual

What about: 

> Virtual Types declared in di.xml files should have a PSR-compliant namespaced name, with `Virtual` as any part of the namespace.

I don't think it's important, if it's a prefix or suffix

**Comment by jissereitsma** [comment]
I agree: The most important thing is that it contains the word `Virtual` in it.

**Comment by shochdoerfer** [comment]
What about enforcing a `\Virtual` postfix similar to the existing `\Proxy` postfix notation? Just to add a bit of consistency...

**Comment by fooman** [comment]
Awesome discussion going on - just depositing a link to the twitter feedback [here](https://twitter.com/jissereitsma/status/993063332784549889)

**Comment by jissereitsma** [comment]
As mentioned in a hangout, I'll (ping @jissereitsma) will investigate whether PHPCS rules are able to inspect XML files at all. If so, let's do it.

**Comment by schmengler** [comment]
Thanks to @lenaorobei for the hint on custom tokenizers: https://github.com/squizlabs/PHP_CodeSniffer/tree/master/src/Tokenizers but I could not find any implementation for XML yet.
